### ISSUE TITLE
Aspire.Npgsql component does not create database

### ISSUE THREAD

**Issue by alexwolfmsft** [dotnet/aspire#1170]
When providing a custom name for a database, the `builder.AddPostgresContainer("pg").AddDatabase("postgresdb")` methods do not seem to create the database in the Postgres container.  For example, with the following code:

```
var postgresdb = builder.AddPostgresContainer("catalog").AddDatabase("catalogdb");

var postgres = builder.AddProject<Projects.AspireSQL>("aspirepostgres")
                       .WithReference(postgresdb);
```

And:
```
builder.AddNpgsqlDataSource("catalogdb");
```

I receive this error:
PostgresException: 3D000: database "catalogdb" does not exist

When I use psql commands in the container to view the list of databases, it looks like this:

<img width="746" alt="image" src="https://github.com/dotnet/aspire/assets/93200798/77653089-50f5-4baf-8f5d-c2ac97e04816">

Should I expect to see a "catalogdb" database listed here?

The connection and query does work if I switch to `AddDatabase("postgres")` to use the default database.

**Comment by DamianEdwards** [comment]
This is by design. The database name is to ensure the appropriate connection information is provided but it's not intended that the database will be auto-created by the hosting APIs given all the variables that can impact database creation.

**Comment by eerhardt** [comment]
> The database name is to ensure the appropriate connection information is provided but it's not intended that the database will be auto-created by the hosting APIs given all the variables that can impact database creation.

I think this is confusing, and not consistent with our other hosting APIs. The first call is "AddPostgresContainer", which creates a Postgres container. The second call is "AddDatabase", which doesn't create a database? That seems wrong that the same verb doesn't do the same thing.

**Comment by alexwolfmsft** [comment]
@eerhardt  It's interesting you mention this because that's the exact reason I assumed it auto generated the db - it autogenerates a redis setup for you.

**Comment by alexwolfmsft** [comment]
> Re-opening to discuss this a bit more. I'm not sure the current behavior is the best we can do.

I know you've seen https://github.com/dotnet/aspire/issues/1168 as well, but just wanted to point out the other db components don't autogenerate the database either - this seems to be a Redis vs Databases inconsistency.

Some .NET tools have historically autogenerated databases for you locally, though I think that was only when using EF.

**Comment by DamianEdwards** [comment]
Redis is more akin to a service than a database, in that you get a Redis instance and connect to that. With databases, you typically get a server which hosts multiple databases, and the database is part of the connection detail. All that said, I'm more than happy to discuss further whether we think the Aspire.Hosting layer should do more to ensure the configured database is actually available.

/Cc @davidfowl @roji for their thoughts.

**Comment by mitchdenny** [comment]
The moment you start auto creating the database you need to start specifying things like collations and that kind of stuff.

The other part of the problem is that you have to figure out where you want to provision the database. If you say you want to do it in the AppHost then suddenly the app host needs to have client drivers for each database which could be dependency hell.

Then you get into lifetime considerations such as the AppHost not being around during deployment so you still need something else to create the database anyway.

Better off to face these considerations up front.

It is possible for a deployment tool like AZD to be enhanced to support creating the database since in the Azure ARM world a database is actually a resource in the resource provider. But this is certainly cloud platform dependent.

**Comment by IEvangelist** [comment]
Perhaps renaming the `AddDatabase` API should be renamed to `AddDatabaseConnection` to make it clearer, that it's not in fact adding a database, but rather a connection to one?

**Comment by mitchdenny** [comment]
LOL ... it is funny how we keep recycling nouns :) We just removed `AddPostgesConnection(...)` ... and now looking to add `AddPostgres(...).AddDatabaseConnection(...)` which does something different.

I've never been super enthused about `AddDatabase(...)` myself so I'm more than open to renaming it. I'm not sure `AddDatabaseConnection(...)` is any better. Perhaps we need something like `SelectDatabase(...)`.

**Comment by DamianEdwards** [comment]
There's a PR out that adds a sample on how to initialize databases when using database containers now: https://github.com/dotnet/aspire-samples/pull/61

**Comment by mitchdenny** [comment]
Flagging for offsite because this has come up a bit.

**Comment by mitchdenny** [comment]
I think we are comfortable with where this is for now. When provisioning you can get the database created for you because the permission set for deployed resources is usually a bit stricter. 

For local development a worker app to manage database creation is our current guidance.

We may revisit this when we look at commanding/tasking.

**Comment by DamianEdwards** [comment]
> This is worse in the case of SqlServer i believe

I'm fairly certain this happens for other database engines too like PostgreSQL.

> So even if i wanted to create the database from code - i can't, the connection string will not work

Entity Framework Core actually handles this specifically in its database creation routines and reconnects to the database with a modified connection string (by stripping off the database name). There's no reason your apps can't do something similar (assuming you're not using EF Core).

> I think it would be even better to create those databases as part of the AppHost configuration though

The resistance to this is that that requires having the database client referenced by the AppHost project and the logic to check for and if not present create the database needs to be specifically implemented for every different database hosting extension (as it's not universal). This is exactly what provider-based ORM libraries like Entity Framework Core exist to do and we're hesitant to reimplement that in the Aspire hosting layer. All that said, I acknowledge that it still might be the right thing to do as this has been a consistent piece of feedback we've had since the first preview.

**Comment by jakoss** [comment]
> Entity Framework Core actually handles this specifically in its database creation routines and reconnects to the database with a modified connection string (by stripping off the database name). There's no reason your apps can't do something similar (assuming you're not using EF Core).

That does explain my issue. I have one service with normal DbContext using postgres and it's working fine. Other service is using SqlServer and it's a Identity Server. I'm trying to set persistence to Identity Server and it's not using DbContext, hence the error i guess.

> The resistance to this is that that requires having the database client referenced by the AppHost project and the logic to check for and if not present create the database needs to be specifically implemented for every different database hosting extension (as it's not universal). This is exactly what provider-based ORM libraries like Entity Framework Core exist to do and we're hesitant to reimplement that in the Aspire hosting layer. All that said, I acknowledge that it still might be the right thing to do as this has been a consistent piece of feedback we've had since the first preview.

Understandable, I guess it's not really that bad. But I'm quite sure it should be documented better, both in the learn.microsoft.com and in the code docs :)

**Comment by holytshirt** [comment]
I just hit this and though I was doing something wrong for hours.  I did not realise the database was not created.
In docker compose 
```
    image: postgres:16-alpine
    environment:
      - POSTGRES_USER=postgres
      - POSTGRES_PASSWORD=postgres
      - POSTGRES_DB=zitadel
```
And database is there ,,,

**Comment by DamianEdwards** [comment]
@holytshirt yes the PostgreSQL container supports changing the name of the default database via the `POSTGRES_DB` environment variable, as does the MySQL container. The [DatabaseContainers sample](https://github.com/dotnet/aspire-samples/blob/d674e52ba5d03755ad3a8e73656cd83995084d25/samples/DatabaseContainers/DatabaseContainers.AppHost/Program.cs#L10) shows this. However not all database containers have this mechanism, including the SQL Server container.

We're still considering what we might do to have databases specified via `AddDatabase` be automatically created if they're not present but haven't settled on a plan as yet.

**Comment by ozzioma** [comment]
Is there a scenario for connecting to an existing database?
That is without provisioning a container for one?

**Comment by DamianEdwards** [comment]
Yes, you can use `builder.AddConnectionString` to model an existing database instance

**Comment by mitchdenny** [comment]
No we provably won't go that direction. We are considering allowing the hosting packages to reference client libraries to support doing things like creating databases for local development.

**Comment by manojlds** [comment]
So how am I supposed to handle this? This is trivial with existing docker-compose setups.

**Comment by eerhardt** [comment]
> This is trivial with existing docker-compose setups.

How do you do it with docker-compose?

**Comment by Matthewsre** [comment]
I encountered the same confusion with `.AddDatabase(...)` not actually creating the database. I see the value in supporting both creating the database and just referencing it.

Here are a few options that make sense to me:

**Option 1: Prefix Conventions (Add, With, ...)**

- Modify `.AddDatabase()` to create the database. This aligns with user expectations where "Add" implies creation, even though it's a breaking change.
- Introduce `.WithDatabase()` (or `.WithDatabaseReference()`) to include database details without creating it, providing clear distinction. Might break expectations around "With" prefix usage for chaining expectations. 
(Any good alternatives for a prefix that makes more sense?)

**Option 2: Suffix Conventions (Reference, ReferenceOnly, ...)**

- Same as Option 1, but introduce `.AddDatabaseReference()` (or `.AddDatabaseReferenceOnly()`). This is really just an alternative name to the proposal by @IEvangelist. This might make sense since the “Add” prefix also seems to have an expected behavior of providing a new output compared to the “With” prefix used for chaining.

**Option 3: Optional Argument (create = true)**

- Add an optional argument to `.AddDatabase(…, bool create = true)`. This avoids breaking changes by defaulting to the current behavior, offering a short-term solution outside major version releases.

Additionally, I would like to see these same considerations made for containers/collections for resources such as CosmosDB and MongoDB.

For context, I use Terraform to build environments (dev, test, ppe, prod), which creates databases and sometimes containers/collections. I'm transitioning to Aspire to simplify dev environment configurations and reduce/remove Terraform. Beyond the dev environment, I plan to use Aspire for generating manifests compatible with Terraform, Pulumi, Bicep, etc. (something similar to one of my examples here: [terraform-azurerm-microservices/examples/exampleB/main.tf at master · Matthewsre/terraform-azurerm-microservices (github.com)](https://github.com/Matthewsre/terraform-azurerm-microservices/blob/master/examples/exampleB/main.tf))

**Comment by rsalus** [comment]
I will admit I also found this confusing when I was starting out, but admittedly I did not have much familiarity with EF core at the time. Looking back, I think this is more of a documentation issue than an architectural one. Covering the basics of actually interacting with the DB (seeding, migrations, etc) when introducing the Aspire functionality would have gone a long way IMO.

**Comment by Alirexaa** [comment]
My proposed API is to introduce `WithDatabaseCreation` in another layer that referenced two projects; `Aspire.Hosting.PostgreSQL` and `Aspire.Npgsql` or `Aspire.Npgsql.EntityFrameworkCore.PostgreSQL`.
It could be more helpful to have this layer to do more things like `WithDatabaseRestore`
and also we could provide settings to completely configure Database properties some collation with `WithDatabaseCreation` as @mitchdenny says in https://github.com/dotnet/aspire/issues/1170#issuecomment-1839514981

**Comment by davidfowl** [comment]
@mitchdenny Is working on infrastructure that will make this possible.

**Comment by DibyodyutiMondal** [comment]
So I wrote this extension method that does what is required but it feels rather hacky:

```csharp
// create the postgres server resource
var fga_pg = builder.AddPostgres(
    "fga-pg",
    parameters.postgres_user,
    parameters.postgres_password,
    builder.Configuration.GetValue<int?>("Parameters:fga-pg-port", null)
  )
  .WithImage("postgres")
  .WithContainerName("fga-pg")
  .WithDataVolume("fga-pg-data")
  .WithLifetime(ContainerLifetime.Persistent);

// create the postgres database resource
// chain the extension function to create the database
var fga_db = fga_pg.AddDatabase("fga-db")
  .WithCreateCommand(true);
```

```csharp
 public static IResourceBuilder<T> WithCreateCommand<T>(
    this IResourceBuilder<T> dbBuilder,
    bool autocreate = false
  ) where T : PostgresDatabaseResource
  {
    var commandName = "create-database";
    dbBuilder.WithCommand(
      commandName,
      "Create Database",
      async ctx =>
      {
        var connectionString = await dbBuilder.Resource
          .ConnectionStringExpression
          .GetValueAsync(ctx.CancellationToken);

        var optionsBuilder = new DbContextOptionsBuilder();
        optionsBuilder.UseNpgsql(connectionString);
        var dbCtx = new DbContext(optionsBuilder.Options);

        await dbCtx.Database.MigrateAsync();

        return CommandResults.Success();
      },
      ctx => ResourceCommandState.Enabled
    );

    if (autocreate)
    {
      dbBuilder.ApplicationBuilder.Eventing.Subscribe<ResourceReadyEvent>(
        // wait for the server to be ready
        dbBuilder.Resource.Parent,
        (ctx, cancel) =>
        {
          var commandAnnotation = dbBuilder.Resource.Annotations
            .OfType<ResourceCommandAnnotation>()
            .SingleOrDefault(a => a.Name == commandName)
            ?? throw new InvalidOperationException("cannot autocreate the database -  the 'create-database' command is not registered'");

          await commandAnnotation.ExecuteCommand.Invoke(new ExecuteCommandContext()
          {
            ServiceProvider = ctx.Services,
            ResourceName = dbBuilder.Resource.Name,
            CancellationToken = cancel
          });
        }
      );
    }

    return dbBuilder;
  }
```

**Comment by DibyodyutiMondal** [comment]
This is something that anyone can use today itself. The drawback is that it has a dependency on EF Core (and Npgsql for Postgres) 

But the extension method can be modifed to support other databases easily - as long as they are supported by EF Core.

But it would be better if we didn't need that. I suppose we can use a barebones sql client to achieve the same for any database that supports SQL. I am not aware of how that will look like, though

**Comment by rbsis** [comment]
A database can simply be created by adding a reference to Aspire.Microsoft.EntityFrameworkCore.SqlServer package and add something like this to the Program .cs of the project that needs the database:

```cs
builder.AddSqlServerDbContext<DbContext>("sql");
builder.Services.BuildServiceProvider().GetRequiredService<DbContext>().Database.EnsureCreated();
```

**Comment by davidfowl** [comment]
In Aspire 9.2 the apphost will create empty databases (and it'll be customizable). See https://github.com/dotnet/aspire/pull/8085
